src/v/utils/process.h

```diff
@@ -0,0 +1,148 @@
+            return "Process does not exist";
+        case process_errc::running:
+            return "Process is already running";
+        case process_errc::non_zero_exit:
+            return "Process exited with non-zero status";
+        case process_errc::signaled:
+            return "Process exited on signal";
+        }
+        return "process::process_errc::unknown";
+    }
+};
+inline const std::error_category& error_category() noexcept {
+    static process_errc_category e;
+    return e;
+}
+inline std::error_code make_error_code(process_errc e) noexcept {
+    return std::error_code(static_cast<int>(e), error_category());
+}
+
+inline process_errc to_process_errc(std::error_code ec) {
+    return process_errc(ec.value());
+}
+
+class process {
+public:
+    process() {}
+    ~process();
+
+    // Spawns a posix fork and starts running the command.
+    //
+    // Output from stdout and stderr are logged at TRACE level.
+    ss::future<std::error_code> spawn(
+      const std::filesystem::path& cmd,
+      ss::experimental::spawn_parameters params);
+
+    // Terminate a running process
+    //
+    // /param: timeout: amount of time to wait for the process to respond to
+    // SIGTERM. Default 1s.
+    template<typename Rep = long long, typename Period = std::ratio<1>>
+    ss::future<std::error_code> terminate(
+      std::chrono::duration<Rep, Period> timeout = std::chrono::seconds(1)) {
+        gate_guard g{_gate};
+        if (!is_running()) {
+            co_return process_errc::does_not_exist;
+        }
+
+        try {
+            _process->terminate();
+        } catch (const std::system_error&) {
+            // The process is already dead
+            co_return process_errc::does_not_exist;
+        }
+
+        auto deadline = ss::steady_clock_type::now() + timeout;
+
+        co_return co_await ss::with_timeout(
+          deadline,
+          ss::do_until(
+            [this, deadline] {
+                // The callback (do_until loop) is not auto canceled on timeout,
+                // so exit when the deadline is exceeded.
+                if (ss::steady_clock_type::now() > deadline) {
+                    return true;
+                }
+
+                return !is_running();
+            },
+            [] { return ss::sleep(std::chrono::milliseconds{5}); }))
+          .then([] {
+              return ss::make_ready_future<process_errc>(process_errc::success);
+          })
+          .handle_exception_type([this](const ss::timed_out_error&) {
+              try {
+                  _process->kill();
+                  return ss::make_ready_future<process_errc>(
+                    process_errc::success);
+              } catch (const std::system_error&) {
+                  // The process is already dead
+                  return ss::make_ready_future<process_errc>(
+                    process_errc::does_not_exist);
+              }
+          });
+    }
+
+    std::pair<int, ss::sstring> exit_status();
```

Comment:
What does this return type mean? Generally I'd suggest named_types or just create a little struct with named members so we know what the string and number represent.

Also documentation for the method might help.

Reply:
Also should this return optional if the process hasn't exited?

Reply:
> > Also should this return optional if the process hasn't exited?
> 
> Returning a negative integer should suffice.

notes from offline sync: I'll do this after I add the factory method because that change will add alot of optionals as well.